File: outbound/retest_t65_t69_20260307.mjs
import fs from 'node:fs';
import path from 'node:path';
import { chromium } from 'playwright';

const repo = '/Users/apple/coding/Eggturtle-breeding-library';
const outbound = path.join(repo, 'outbound');
fs.mkdirSync(outbound, { recursive: true });

function readEnv(filePath) {
  const env = {};
  for (const line of fs.readFileSync(filePath, 'utf8').split(/\r?\n/)) {
    if (!line || line.trim().startsWith('#')) continue;
    const idx = line.indexOf('=');
    if (idx === -1) continue;
    const key = line.slice(0, idx).trim();
    let value = line.slice(idx + 1).trim();
    if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
      value = value.slice(1, -1);
    }
    env[key] = value;
  }
  return env;
}

const env = readEnv(path.join(repo, '.env'));
const account = env.TENANT_LOGIN_ACCOUNT;
const password = env.TENANT_LOGIN_PASSWORD;
const baseUrl = 'http://127.0.0.1:30010';
const apiBase = 'http://127.0.0.1:30011';
const code = `QA-${Date.now().toString().slice(-6)}`;

const result = {
  startedAt: new Date().toISOString(),
  tenantSlug: null,
  productId: null,
  shareToken: null,
  screenshots: [],
  consoleErrors: [],
  pageErrors: [],
  checks: [],
};

function add(level, title, detail) {
  result.checks.push({ level, title, detail });
}

async function shot(page, name) {
  const file = path.join(outbound, name);
  await page.screenshot({ path: file, fullPage: true });
  result.screenshots.push(path.relative(repo, file));
}

async function loginViaApi(request) {
  const resp = await request.post(`${apiBase}/auth/password-login`, { data: { login: account, password } });
  if (!resp.ok()) throw new Error(`password-login failed: ${resp.status()} ${await resp.text()}`);
  const json = await resp.json();
  const accessToken = json.accessToken;
  const meResp = await request.get(`${apiBase}/tenants/me`, { headers: { Authorization: `Bearer ${accessToken}` } });
  if (!meResp.ok()) throw new Error(`tenants/me failed: ${meResp.status()} ${await meResp.text()}`);
  const me = await meResp.json();
  const tenantSlug = me.tenants?.[0]?.tenant?.slug;
  if (!tenantSlug) throw new Error('No tenant slug found');
  const switchResp = await request.post(`${apiBase}/auth/switch-tenant`, {
    headers: { Authorization: `Bearer ${accessToken}` },
    data: { slug: tenantSlug },
  });
  if (!switchResp.ok()) throw new Error(`switch-tenant failed: ${switchResp.status()} ${await switchResp.text()}`);
  const switched = await switchResp.json();
  return { tenantSlug, accessToken: switched.accessToken };
}

function parseHexColor(value) {
  const input = (value || '').trim().toLowerCase();
  if (/^#[0-9a-f]{6}$/.test(input)) return input;
  const match = input.match(/rgba?\((\d+),\s*(\d+),\s*(\d+)/i);
  if (!match) return input;
  return '#' + match.slice(1, 4).map((n) => Number(n).toString(16).padStart(2, '0')).join('');
}

const browser = await chromium.launch({ headless: true });
const context = await browser.newContext({ viewport: { width: 1440, height: 1200 } });
const page = await context.newPage();
page.on('console', (msg) => { if (msg.type() === 'error') result.consoleErrors.push(msg.text()); });
page.on('pageerror', (err) => { result.pageErrors.push(String(err)); });

try {
  const auth = await loginViaApi(context.request);
  result.tenantSlug = auth.tenantSlug;

  await page.goto(`${baseUrl}/login`, { waitUntil: 'networkidle' });
  await page.evaluate((token) => window.localStorage.setItem('eggturtle.accessToken', token), auth.accessToken);

  await page.goto(`${baseUrl}/app/${auth.tenantSlug}/products`, { waitUntil: 'networkidle' });
  await page.getByRole('button', { name: '新建乌龟' }).click();
  await page.waitForTimeout(600);
  await shot(page, 'retest-t65-create-drawer-default.png');

  const seriesSelect = page.locator('#create-drawer-series');
  const newSeriesButton = page.getByRole('button', { name: '新建系列' });
  const seriesVisible = await seriesSelect.isVisible();
  const buttonVisible = await newSeriesButton.isVisible();
  const selectBox = await seriesSelect.boundingBox();
  const buttonBox = await newSeriesButton.boundingBox();
  add(seriesVisible && buttonVisible ? 'PASS' : 'FAIL', '新建乌龟抽屉系列区可用', `seriesVisible=${seriesVisible}; buttonVisible=${buttonVisible}; selectBox=${JSON.stringify(selectBox)}; buttonBox=${JSON.stringify(buttonBox)}`);

  await newSeriesButton.click();
  await page.waitForTimeout(300);
  await shot(page, 'retest-t65-create-drawer-new-series.png');
  await page.getByPlaceholder('系列名称（必填）').fill('白化');
  await page.getByLabel('产品编码（必填）').fill(code);
  await page.getByRole('button', { name: /^确认创建/ }).click();
  await page.waitForLoadState('networkidle');
  await page.waitForTimeout(1200);
  await shot(page, 'retest-t65-product-list-after-create.png');

  const card = page.getByText(code).first();
  await card.waitFor({ timeout: 10000 });
  const cardContainer = card.locator('xpath=ancestor::*[self::article or self::a][1]');
  const cardText = await cardContainer.innerText().catch(async () => await page.locator('body').innerText());
  add(/白化/.test(cardText) ? 'PASS' : 'FAIL', '列表页显示新系列名称', `code=${code}; cardText=${cardText.slice(0, 500)}`);
  add(/NEW-SERIES|\bNEW\b/.test(cardText) ? 'FAIL' : 'PASS', '列表页未暴露 NEW-SERIES/NEW', `code=${code}; cardText=${cardText.slice(0, 500)}`);

  const productResp = await context.request.get(`${apiBase}/products?keyword=${encodeURIComponent(code)}`, {
    headers: { Authorization: `Bearer ${auth.accessToken}` },
  });
  if (!productResp.ok()) throw new Error(`product lookup failed: ${productResp.status()} ${await productResp.text()}`);
  const productJson = await productResp.json();
  const created = (productJson.items || []).find((item) => item.code === code) || productJson.items?.[0];
  if (!created?.id) throw new Error('Created product not found via API');
  result.productId = created.id;

  const meResp = await context.request.get(`${apiBase}/me`, { headers: { Authorization: `Bearer ${auth.accessToken}` } });
  if (!meResp.ok()) throw new Error(`me failed: ${meResp.status()} ${await meResp.text()}`);
  const meJson = await meResp.json();
  const tenantId = meJson.tenantId;
  if (!tenantId) throw new Error('No tenantId in /me response');

  const shareResp = await context.request.post(`${apiBase}/shares`, {
    headers: { Authorization: `Bearer ${auth.accessToken}` },
    data: { resourceType: 'tenant_feed', resourceId: tenantId },
  });
  if (!shareResp.ok()) throw new Error(`share create failed: ${shareResp.status()} ${await shareResp.text()}`);
  const shareJson = await shareResp.json();
  result.shareToken = shareJson.share.shareToken;

  await page.goto(`${baseUrl}/public/s/${result.shareToken}/products/${created.id}`, { waitUntil: 'networkidle' });
  await page.waitForTimeout(1200);
  await shot(page, 'retest-t65-public-product-detail.png');
  const detailText = await page.locator('body').innerText();
  add(/白化/.test(detailText) ? 'PASS' : 'FAIL', '详情页显示新系列名称', detailText.slice(0, 800));
  add(/NEW-SERIES|\bNEW\b/.test(detailText) ? 'FAIL' : 'PASS', '详情页未暴露 NEW-SERIES/NEW', detailText.slice(0, 800));

  await page.goto(`${baseUrl}/app/${auth.tenantSlug}/share-presentation`, { waitUntil: 'networkidle' });
  await page.waitForTimeout(1200);
  await shot(page, 'retest-t69-share-presentation-initial.png');

  const customColorLabelCount = await page.getByText('自定义颜色', { exact: true }).count();
  add(customColorLabelCount > 0 ? 'FAIL' : 'PASS', '分享配置不应再暴露自定义颜色入口', `customColorLabelCount=${customColorLabelCount}`);

  const colorButtons = page.locator('button[aria-pressed]');
  const pressedCount = await colorButtons.count();
  const pressedSummaries = [];
  for (let i = 0; i < pressedCount; i += 1) {
    const btn = colorButtons.nth(i);
    const pressed = await btn.getAttribute('aria-pressed');
    if (pressed === 'true') {
      const text = (await btn.innerText()).trim();
      const borderColor = parseHexColor(await btn.evaluate((el) => getComputedStyle(el).borderColor));
      const bgColor = parseHexColor(await btn.evaluate((el) => getComputedStyle(el).backgroundColor));
      pressedSummaries.push({ text, borderColor, bgColor });
    }
  }
  const hasYellowSelected = pressedSummaries.some((item) => item.text.includes('金黄') || item.borderColor === '#ffd400' || item.bgColor === '#fff7d0');
  add(hasYellowSelected ? 'PASS' : 'FAIL', '默认选中态应为黄色', JSON.stringify(pressedSummaries));

  const colorInput = page.locator('input[aria-label="主题主色 自定义颜色"]');
  if (await colorInput.count()) {
    await colorInput.fill('#06b6d4');
    await page.getByRole('button', { name: '保存并立即生效' }).click();
    await page.waitForTimeout(1200);
    await page.getByRole('button', { name: '金黄' }).click();
    await page.getByRole('button', { name: '保存并立即生效' }).click();
    await page.waitForTimeout(1200);
    await shot(page, 'retest-t69-share-presentation-after-reset-to-yellow.png');
    const currentColorTextCount = await page.getByText('当前色', { exact: true }).count();
    add(currentColorTextCount > 0 ? 'FAIL' : 'PASS', '取消历史自定义颜色后不应残留“当前色”', `currentColorTextCount=${currentColorTextCount}`);
  }
} finally {
  await browser.close();
}

result.finishedAt = new Date().toISOString();
result.conclusion = result.checks.some((item) => item.level === 'FAIL') ? '不可 push' : '可 push';
fs.writeFileSync(path.join(outbound, 'retest-t65-t69-20260307-result.json'), JSON.stringify(result, null, 2));
const lines = [
  '# Retest T65/T69',
  '',
  `- time: ${result.finishedAt}`,
  `- tenant: ${result.tenantSlug}`,
  `- productId: ${result.productId ?? 'n/a'}`,
  `- shareToken: ${result.shareToken ?? 'n/a'}`,
  `- conclusion: ${result.conclusion}`,
  '',
  '## Screenshots',
  ...result.screenshots.map((item) => `- ${item}`),
  '',
  '## Checks',
  ...result.checks.map((item) => `- [${item.level}] ${item.title}: ${item.detail}`),
  '',
  `- consoleErrors: ${result.consoleErrors.length}`,
  `- pageErrors: ${result.pageErrors.length}`,
  '',
];
fs.writeFileSync(path.join(outbound, 'retest-t65-t69-20260307-report.md'), lines.join('\n'));
console.log(lines.join('\n'));

File: outbound/retest_t65_t69_20260307.py
from pathlib import Path
from datetime import datetime
import json
import re
from playwright.sync_api import sync_playwright

repo = Path('/Users/apple/coding/Eggturtle-breeding-library')
outbound = repo / 'outbound'
outbound.mkdir(parents=True, exist_ok=True)

def read_env(path: Path):
    data = {}
    for raw in path.read_text().splitlines():
        line = raw.strip()
        if not line or line.startswith('#') or '=' not in raw:
            continue
        key, value = raw.split('=', 1)
        value = value.strip().strip('"').strip("'")
        data[key.strip()] = value
    return data

def parse_hex_color(value: str) -> str:
    s = (value or '').strip().lower()
    if re.fullmatch(r'#[0-9a-f]{6}', s):
        return s
    m = re.search(r'rgba?\((\d+),\s*(\d+),\s*(\d+)', s)
    if not m:
        return s
    return '#' + ''.join(f'{int(part):02x}' for part in m.groups())

env = read_env(repo / '.env')
account = env['TENANT_LOGIN_ACCOUNT']
password = env['TENANT_LOGIN_PASSWORD']
base_url = 'http://127.0.0.1:30010'
api_base = 'http://127.0.0.1:30011'
code = f"ZZ{str(int(datetime.now().timestamp() * 1000))[-8:]}"

result = {
    'startedAt': datetime.now().isoformat(),
    'tenantSlug': None,
    'productId': None,
    'shareToken': None,
    'seriesId': None,
    'seriesCode': None,
    'seriesName': None,
    'screenshots': [],
    'consoleErrors': [],
    'pageErrors': [],
    'checks': [],
    'health': {},
}

def add(level, title, detail):
    result['checks'].append({'level': level, 'title': title, 'detail': detail})

def shot(page, filename):
    path = outbound / filename
    page.screenshot(path=str(path), full_page=True)
    result['screenshots'].append(str(path.relative_to(repo)))

with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)
    context = browser.new_context(viewport={'width': 1440, 'height': 1200})
    page = context.new_page()
    page.on('console', lambda msg: result['consoleErrors'].append(msg.text) if msg.type == 'error' else None)
    page.on('pageerror', lambda err: result['pageErrors'].append(str(err)))

    health_resp = context.request.get(f'{api_base}/health')
    result['health']['api'] = health_resp.status
    add('PASS' if health_resp.ok else 'FAIL', '本地 API 健康检查', f'status={health_resp.status}')

    login_resp = context.request.post(f'{api_base}/auth/password-login', data={'login': account, 'password': password})
    assert login_resp.ok, login_resp.text()
    access_token = login_resp.json()['accessToken']

    tenants_resp = context.request.get(f'{api_base}/tenants/me', headers={'Authorization': f'Bearer {access_token}'})
    assert tenants_resp.ok, tenants_resp.text()
    tenant_slug = tenants_resp.json()['tenants'][0]['tenant']['slug']
    result['tenantSlug'] = tenant_slug

    switch_resp = context.request.post(
        f'{api_base}/auth/switch-tenant',
        headers={'Authorization': f'Bearer {access_token}'},
        data={'slug': tenant_slug},
    )
    assert switch_resp.ok, switch_resp.text()
    switched_token = switch_resp.json()['accessToken']

    page.goto(f'{base_url}/login', wait_until='networkidle')
    page.evaluate("token => window.localStorage.setItem('eggturtle.accessToken', token)", switched_token)

    page.goto(f'{base_url}/app/{tenant_slug}/products', wait_until='networkidle')
    page.get_by_role('button', name='新建产品').click()
    page.wait_for_timeout(700)
    shot(page, 'retest-t65-create-drawer-default.png')

    series_select = page.locator('#create-drawer-series')
    new_series_button = page.get_by_role('button', name='新建系列')
    add(
        'PASS' if (series_select.is_visible() and new_series_button.is_visible()) else 'FAIL',
        '新建乌龟系列下拉 + 新建系列按钮是否顺手',
        f'selectBox={series_select.bounding_box()}; buttonBox={new_series_button.bounding_box()}',
    )

    new_series_button.click()
    page.wait_for_timeout(300)
    shot(page, 'retest-t65-create-drawer-new-series.png')
    page.locator('#create-drawer-code').fill(code)
    page.get_by_placeholder('系列名称（必填）').fill('白化')
    page.get_by_role('button', name='确认创建').click()
    page.wait_for_timeout(1800)

    body_after_create = page.locator('body').inner_text()
    if '创建成功' in body_after_create:
        add('PASS', '新建乌龟提交成功', body_after_create[:800])
    else:
        add('FAIL', '新建乌龟提交未返回成功态', body_after_create[:1200])

    product_resp = context.request.get(f'{api_base}/products?page=1&pageSize=50', headers={'Authorization': f'Bearer {switched_token}'})
    assert product_resp.ok, product_resp.text()
    products = product_resp.json().get('products', [])
    created = next((item for item in products if item.get('code') == code), None)
    assert created and created.get('id'), f'Created product not found via API for code={code}'
    result['productId'] = created['id']

    series_resp = context.request.get(f'{api_base}/series?page=1&pageSize=50', headers={'Authorization': f'Bearer {switched_token}'})
    assert series_resp.ok, series_resp.text()
    series_items = series_resp.json().get('items', [])
    series_map = {item['id']: item for item in series_items}
    created_series = series_map.get(created['seriesId'])
    result['seriesId'] = created.get('seriesId')
    result['seriesCode'] = created_series.get('code') if created_series else None
    result['seriesName'] = created_series.get('name') if created_series else None

    if created_series:
        add('PASS' if created_series.get('name') == '白化' else 'FAIL', '新建系列名称回显为白化', json.dumps(created_series, ensure_ascii=False))
        add('PASS' if created_series.get('code') == '白化' else 'FAIL', '新建系列编码未异常回显为产品编码/内部值', json.dumps(created_series, ensure_ascii=False))
        add('FAIL' if re.search(r'NEW-SERIES|\bNEW\b', json.dumps(created_series, ensure_ascii=False)) else 'PASS', '新建系列未回退成 NEW-SERIES/NEW', json.dumps(created_series, ensure_ascii=False))
    else:
        add('FAIL', '无法通过 API 找到新建系列', f'product={json.dumps(created, ensure_ascii=False)}')

    page.goto(f'{base_url}/app/{tenant_slug}/series', wait_until='networkidle')
    page.wait_for_timeout(1200)
    shot(page, 'retest-t65-series-page.png')
    series_page_text = page.locator('body').inner_text()
    add('PASS' if '白化' in series_page_text else 'FAIL', '系列页面可见白化', series_page_text[:1500])
    add('FAIL' if 'ID ' in series_page_text else 'PASS', '系列页面不应暴露内部 ID', series_page_text[:1500])
    add('FAIL' if result['seriesCode'] == code else 'PASS', '系列编码不应错误等于产品编码', f'productCode={code}; seriesCode={result["seriesCode"]}; seriesName={result["seriesName"]}')

    me_resp = context.request.get(f'{api_base}/me', headers={'Authorization': f'Bearer {switched_token}'})
    assert me_resp.ok, me_resp.text()
    tenant_id = me_resp.json()['tenantId']
    share_resp = context.request.post(
        f'{api_base}/shares',
        headers={'Authorization': f'Bearer {switched_token}'},
        data={'resourceType': 'tenant_feed', 'resourceId': tenant_id},
    )
    assert share_resp.ok, share_resp.text()
    result['shareToken'] = share_resp.json()['share']['shareToken']

    page.goto(f'{base_url}/app/{tenant_slug}/share-presentation', wait_until='networkidle')
    page.wait_for_timeout(1200)
    shot(page, 'retest-t69-share-presentation-initial.png')

    custom_color_count = page.get_by_text('自定义颜色', exact=True).count()
    add('FAIL' if custom_color_count > 0 else 'PASS', '分享配置取消自定义颜色后默认选中态是否为黄色 - 不应再有“自定义颜色”入口', f'customColorLabelCount={custom_color_count}')

    pressed_summaries = []
    color_buttons = page.locator('button[aria-pressed]')
    for i in range(color_buttons.count()):
        btn = color_buttons.nth(i)
        if btn.get_attribute('aria-pressed') == 'true':
            pressed_summaries.append({
                'text': btn.inner_text().strip(),
                'borderColor': parse_hex_color(btn.evaluate('(el) => getComputedStyle(el).borderColor')),
                'bgColor': parse_hex_color(btn.evaluate('(el) => getComputedStyle(el).backgroundColor')),
            })
    has_yellow_selected = any(item['text'].find('金黄') >= 0 or item['borderColor'] == '#ffd400' or item['bgColor'] == '#fff7d0' for item in pressed_summaries)
    add('PASS' if has_yellow_selected else 'FAIL', '分享配置默认选中态为黄色', json.dumps(pressed_summaries, ensure_ascii=False))

    color_input = page.locator('input[aria-label="主题主色 自定义颜色"]')
    if color_input.count() > 0:
        color_input.fill('#06b6d4')
        page.get_by_role('button', name='保存并立即生效').click()
        page.wait_for_timeout(1200)
        page.get_by_role('button', name='金黄').click()
        page.get_by_role('button', name='保存并立即生效').click()
        page.wait_for_timeout(1200)
        shot(page, 'retest-t69-share-presentation-after-reset-to-yellow.png')
        current_color_count = page.get_by_text('当前色', exact=True).count()
        add('FAIL' if current_color_count > 0 else 'PASS', '历史自定义颜色不应残留为“当前色”', f'currentColorTextCount={current_color_count}')

    browser.close()

result['finishedAt'] = datetime.now().isoformat()
result['conclusion'] = '不可 push' if any(item['level'] == 'FAIL' for item in result['checks']) else '可 push'
(outbound / 'retest-t65-t69-20260307-result.json').write_text(json.dumps(result, ensure_ascii=False, indent=2))
report = [
    '# Retest T65/T69',
    '',
    f"- time: {result['finishedAt']}",
    f"- tenant: {result['tenantSlug']}",
    f"- productId: {result['productId']}",
    f"- seriesId: {result['seriesId']}",
    f"- seriesCode: {result['seriesCode']}",
    f"- seriesName: {result['seriesName']}",
    f"- shareToken: {result['shareToken']}",
    f"- conclusion: {result['conclusion']}",
    '',
    '## Screenshots',
    *[f'- {item}' for item in result['screenshots']],
    '',
    '## Checks',
    *[f"- [{item['level']}] {item['title']}: {item['detail']}" for item in result['checks']],
    '',
    f"- consoleErrors: {len(result['consoleErrors'])}",
    f"- pageErrors: {len(result['pageErrors'])}",
    '',
]
(outbound / 'retest-t65-t69-20260307-report.md').write_text('\n'.join(report))
print('\n'.join(report))
